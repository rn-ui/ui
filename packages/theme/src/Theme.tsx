/* tslint:disable */
import deepMerge from './tools/deepMerge';
import mergeComponentAndThemes from './tools/mergeComponentAndThemes';
import { resolveIncludes } from './tools/resolveIncludes';
import normalizeTheme from './normalizer/normalizeTheme';
import defaultTheme from './resources/default';

interface Themes {
  default: object;
  light: object;
  dark: object;
}

// @ts-ignore
const themes: Themes = {};

// Privates, ideally those should be symbols
const THEME_STYLES = '@southem.theme';

const resolveStyle = (style, baseStyle) => normalizeTheme(resolveIncludes(style, baseStyle));

/**
 * Theme
 *
 * @example
 * const defaultTheme = {};
 * const defaultThemePro = {};
 *
 * Theme.registerTheme('default', [
 *   defaultTheme,
 *   defaultThemePro,
 * ]);
 */
class Theme {

  constructor(themeStyle) {
    // @ts-ignore
    // this[THEME_STYLES] = resolveStyle(themeStyle);
  }

  /**
   * registerTheme
   * @param {String} themeName
   * @param {Array|Object} styleSheets
   */
  static registerTheme(themeName, styleSheets: Array<Object>|Object): void {
    let mergedStyleSheets = styleSheets;

    if (Array.isArray(styleSheets)) {
      mergedStyleSheets = deepMerge({}, ...styleSheets);
    }

    // themes[themeName] = new Theme(themeName, mergedStyleSheets);
    // @ts-ignore
    themes[themeName] = resolveStyle(mergedStyleSheets);
  }

  /**
   * registerDefaultTheme
   * @param {Object} styleSheets
   */
  static registerDefaultTheme(styleSheets): void {
    Theme.registerTheme('default', styleSheets);
  }

  /**
   * getTheme
   * @param {String} themeName
   * @return {Object}
   */
  static getTheme(themeName): Object {
    // @ts-ignore
    if (themeName === 'default' && !themes.default) {
      Theme.registerDefaultTheme(defaultTheme);
    }

    return themes[themeName];
  }

  /**
   * getDefaultTheme
   * @return {Object}
   */
  static getDefaultTheme(): Object {
    return Theme.getTheme('default');
  }

  /**
   * isDefine
   * @param {String} themeName
   * @return {Boolean}
   */
  static isDefine(themeName): Boolean {
    return !!themes[themeName];
  }
}

export {
  Theme as default,
  themes,
};
