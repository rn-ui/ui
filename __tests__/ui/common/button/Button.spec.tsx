import React from 'react';
import {
  Image,
  ImageProps,
  TouchableOpacity,
  ImageSourcePropType,
} from 'react-native';
import {
  render,
  shallow,
  fireEvent,
  RenderAPI,
} from 'react-native-testing-library';
import Theme, {ThemeProvider} from '@southem/theme';
import {
  darkTheme,
  Button,
  ButtonProps,
} from '@southem/ui';

Theme.registerDefaultTheme(darkTheme);

const Mock = (props?: ButtonProps): React.ReactElement<{}> => {
  return (
    <ThemeProvider
      theme={'default'}>
      <Button {...props} />
    </ThemeProvider>
  );
};

const renderComponent = (props?: ButtonProps): RenderAPI => {
  return render(
    <Mock {...props} />,
  );
};

describe('@button: matches snapshot', () => {

  describe('* interaction', () => {
    it('* stateless', () => {
      const component: RenderAPI = renderComponent();
      const {output} = shallow(component.getByType(Button));

      expect(output).toMatchSnapshot();
    });
  });

  describe('* appearance', () => {

    const iconSource: ImageSourcePropType = {uri: 'https://akveo.github.io/eva-icons/fill/png/128/star.png'};

    const icon = (style): React.ReactElement<ImageProps> => {
      return (
        <Image
          source={iconSource}
          style={style}
        />
      );
    };

    const text: React.ReactText = 'BUTTON';

    it('* empty', () => {
      const component: RenderAPI = renderComponent();
      const {output} = shallow(component.getByType(Button));

      expect(output).toMatchSnapshot();
    });

    it('* icon', () => {
      // @ts-ignore
      const component: RenderAPI = renderComponent({
        icon,
      });
      const {output} = shallow(component.getByType(Button));

      expect(output).toMatchSnapshot();
    });

    it('* text', () => {
      // @ts-ignore
      const component: RenderAPI = renderComponent({
        children: text,
      });
      const {output} = shallow(component.getByType(Button));

      expect(output).toMatchSnapshot();
    });

    it('* icon and text', () => {
      // @ts-ignore
      const component: RenderAPI = renderComponent({
        icon,
        children: text,
      });
      const {output} = shallow(component.getByType(Button));

      expect(output).toMatchSnapshot();
    });

    it('* icon and text (styled)', () => {
      // @ts-ignore
      const component: RenderAPI = renderComponent({
        icon,
        children: text,
        size: 'big',
        textStyle: {
          fontSize: 32,
          lineHeight: 34,
        },
      });
      const {output} = shallow(component.getByType(Button));

      expect(output).toMatchSnapshot();
    });
  });
});

describe('@button: component checks', () => {

  it('* emits onPress', () => {
    const onPress = jest.fn();

    // @ts-ignore
    const component: RenderAPI = renderComponent({
      onPress,
    });

    fireEvent.press(component.getByType(TouchableOpacity));

    expect(onPress).toBeCalled();
  });
});
