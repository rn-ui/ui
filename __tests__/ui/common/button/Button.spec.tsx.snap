// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@button: matches snapshot * appearance * empty 1`] = `
<@theme(Animated(Button))
  animationOptions={Object {}}
  forwardedRef={null}
/>
`;

exports[`@button: matches snapshot * appearance * icon 1`] = `
<@theme(Animated(Button))
  animationOptions={Object {}}
  forwardedRef={null}
  icon={[Function]}
/>
`;

exports[`@button: matches snapshot * appearance * icon and text (styled) 1`] = `
<@theme(Animated(Button))
  animationOptions={Object {}}
  forwardedRef={null}
  icon={[Function]}
  size="big"
  textStyle={
    Object {
      "fontSize": 32,
      "lineHeight": 34,
    }
  }
>
  BUTTON
</@theme(Animated(Button))>
`;

exports[`@button: matches snapshot * appearance * icon and text 1`] = `
<@theme(Animated(Button))
  animationOptions={Object {}}
  forwardedRef={null}
  icon={[Function]}
>
  BUTTON
</@theme(Animated(Button))>
`;

exports[`@button: matches snapshot * appearance * text 1`] = `
<@theme(Animated(Button))
  animationOptions={Object {}}
  forwardedRef={null}
>
  BUTTON
</@theme(Animated(Button))>
`;

exports[`@button: matches snapshot * interaction * stateless 1`] = `
<@theme(Animated(Button))
  animationOptions={Object {}}
  forwardedRef={null}
/>
`;
